lib/test_highlevel.ml

```diff
@@ -12,78 +12,100 @@
+    type input_t
+    (* The type of outputs from a system being tested. *)
+    type output_t
+
+    (** Alcotest needs to be told what it is testing.
+        			* The module Alcotest_comparators provides useful combinators to create
+        			  these values. *)
+    val alco_testable : output_t Alcotest.testable
+
+    (* Helper functions for printing error messages on test failure. *)
+    val string_of_input_t : input_t -> string
+    val string_of_output_t : output_t -> string
+  end
+
+  module type STATE = sig
+    (* The type of system state, which will be modified by inputs to the system. *)
+    type state_t
+    (* Create a base system state. *)
+    val create_default_state : unit -> state_t
+  end
+
+  module type STATELESS_TEST = sig
+    module Io : IO
+    (* A function to transform an input into an output. *)
+    val transform : Io.input_t -> Io.output_t
+    (* A list of input/output pairs. *)
+    val tests : [> `Documented of (string * Alcotest.speed_level * Io.input_t  * Io.output_t) list
+                |  `QuickAndAutoDocumented of (Io.input_t * Io.output_t) list]
+  end
+
+  module type STATEFUL_TEST = sig
+    module Io : IO
+    module State : STATE
+    (* A function to apply an input to the system state. *)
+    val load_input : State.state_t -> Io.input_t -> unit
+    (* A function to extract an output from the system state. How this is done
+       		 * may depend on the input to the test. *)
+    val extract_output : State.state_t -> Io.input_t -> Io.output_t
+    (* A list of input/output pairs. *)
+    val tests : [> `Documented of (string * Alcotest.speed_level * Io.input_t  * Io.output_t) list
+                |  `QuickAndAutoDocumented of (Io.input_t * Io.output_t) list]
+  end
+
+  (* Turn a stateful test module into a stateless test module. *)
+  module EncapsulateState(T: STATEFUL_TEST) = struct
+    module Io = T.Io
+
+    let transform input =
+      let state = T.State.create_default_state () in
+      T.load_input state input;
+      T.extract_output state input
+
+    let tests = T.tests
+  end
+
+  module MakeStateless(T: STATELESS_TEST) : sig
+    val tests : unit Alcotest.test_case list
+  end
+  = struct
+    let title input expected_output = Printf.sprintf "%s -> %s"
+        (T.Io.string_of_input_t input)
+        (T.Io.string_of_output_t expected_output) |> Xstringext.String.strip (fun c -> c == '\n')
+
+    let prune_if_too_long s idx =
+      let max_size = 150 in
+      let len = String.length s in
+      if len <= max_size then
+        s
+      else
+        let template = Format.sprintf "%d: %s...%s" idx in
+        let slice_len = (max_size - String.length (template "" "")) / 2 in
+        template (String.sub s 0 slice_len) (String.sub s (len - slice_len - 1) slice_len)
+
+    let test_equal ~input ~expected_output () =
+      Alcotest.check T.Io.alco_testable (title input expected_output) (T.transform input) expected_output
+
+    let tests =	match T.tests with
```

Comment:
Better put the `match` on the next line.

Reply:
i've removed the sneaky tab